netdata/netdata
# Improve PID monitoring (step 2)

##### Summary
Fixes https://github.com/netdata/netdata/issues/12343

This PR is reducing CPU and memory usage for eBPF.plugin.
It is also disabling integration by default integration for legacy kernels, and some kernels that are LTS, but they do not have `trampoline`.
This PR is also bringing `co-re` codes for more threads.

##### Test Plan

1. Compile this branch.
2. Start netdata using the default configuration for `ebpf.plugin` and verify that is your kernel was compiled with `CONFIG_DEBUG_INFO_BTF` (zgrep BTF /proc/config.gz) the integration with `apps` and `cgroup` will be enabled by default, on the other hand, you should not have them enabled. 
3. Verify that you do not have `libbpf` errors (grep libbpf /var/log/netdata/error.log)  for threads different of `socket`. For `socket` this is expected after kernel `5.19` to be  released. This will be fixed in next PR.
4. Now stop netdata and enabling all threads, you should have charts normally.
5. Veriify that memory used by `ebpf.programs` were reduced, they should have `10992` instead the traditional `32768` by default when you run `bpftool map show`.

##### Additional Information
This PR was tested on:

| Linux Distribution | kernel version | Threads     | bpftool |
|--------------------|----------------|-------------|-----------|
| Slackware Current  |     5.19.2    | [slackware_5_19_threads.txt](https://github.com/netdata/netdata/files/9388224/slackware_5_19_threads.txt) | [slackware_5_19_bpftool.txt](https://github.com/netdata/netdata/files/9388223/slackware_5_19_bpftool.txt) |
| Arch Linux         |  5.19.2 | [arch_5_19_threads.txt](https://github.com/netdata/netdata/files/9388226/arch_5_19_threads.txt) | [arch_5_19_bpftool.txt](https://github.com/netdata/netdata/files/9388225/arch_5_19_bpftool.txt)|
| Ubuntu 22.04       | 5.15.0-33-generic   | [ubuntu_5_15_threads.txt](https://github.com/netdata/netdata/files/9388229/ubuntu_5_15_threads.txt) | [ubuntu_5_15_bpftool.txt](https://github.com/netdata/netdata/files/9388228/ubuntu_5_15_bpftool.txt) |
| Alma 9             | 5.14.0-70.17.1.el9_0.x86_64 | [alma_5_14_threads.txt](https://github.com/netdata/netdata/files/9388230/alma_5_14_threads.txt) |  --   |
| Debian 11          | 5.10.0-16-amd64 | [debian_5_10_threads.txt](https://github.com/netdata/netdata/files/9388232/debian_5_10_threads.txt) | [debian_5_10_bpftool.txt](https://github.com/netdata/netdata/files/9388231/debian_5_10_bpftool.txt) |
| Alma 8.6           | 4.18.0-372.19.1.el8_6 | [alma_4_18_threads.txt](https://github.com/netdata/netdata/files/9388236/alma_4_18_threads.txt) | [alma8_4_18_bpftool.txt](https://github.com/netdata/netdata/files/9388235/alma8_4_18_bpftool.txt) |
| Ubuntu 18.04       |   4.15.0-180   | [ubuntu_4_15_threads.txt](https://github.com/netdata/netdata/files/9388239/ubuntu_4_15_threads.txt) | [ubuntu_4_15_bpftool.txt](https://github.com/netdata/netdata/files/9388238/ubuntu_4_15_bpftool.txt) |

<details> <summary>For users: How does this change affect me?</summary>
Describe the PR affects users: 
- Which area of Netdata is affected by the change? ebpf.plugin
- Can they see the change or is it an under the hood? If they can see it, where? The plugin will use less resources from host.
- How is the user impacted by the change?  A better performance from plugin;
- What are there any benefits of the change? plugin improvement.
</details>
